In [1]:
import numpy as np 
import scipy as sc
import pandas as pd

# Exploración del DataFrame

En este notebook, utilizaremos algunas herramientas de Pandas para explorar el registro simultáneo de las células. Para esto, primero cargaremos el archivo preprocesado en el Notebook anterior:

In [2]:
rawdata = pd.read_csv('./CellsData_pd.csv',index_col=0)
nRows = rawdata.shape[0]
nCells = rawdata.shape[1]
rawdata.head()

cell_1  cell_2  cell_3  cell_4  cell_5  cell_6  cell_7  cell_8  cell_9  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   cell_10  ...  cell_29  cell_30  cell_31  cell_32  cell_33  cell_34  \
0        0  ...        0        0        0        0        0        0   
1        0  ...        0        0        0        0        0        0   
2        0  ...        0        0        0        0        0        0   
3        1  ...        0        0        0        0        0        0   
4        1  ...        0        0        0        0        0        0   

   cell_35  cell_36  cell_37  cell_38  
0        0        0        0        0  
1        0        0        0        0  
2        0        0        0        0  
3        0        0        0        0  
4        0        0        0        0  

[5 rows x 38 columns]

# 1. Información general del DataFrame
## 1.1 Método .info( )
Para acceder a la información general del DataFrame, utilizamos el método **.info( )**:

In [3]:
rawdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1960 entries, 0 to 1959
Data columns (total 38 columns):
cell_1     1960 non-null int64
cell_2     1960 non-null int64
cell_3     1960 non-null int64
cell_4     1960 non-null int64
cell_5     1960 non-null int64
cell_6     1960 non-null int64
cell_7     1960 non-null int64
cell_8     1960 non-null int64
cell_9     1960 non-null int64
cell_10    1960 non-null int64
cell_11    1960 non-null int64
cell_12    1960 non-null int64
cell_13    1960 non-null int64
cell_14    1960 non-null int64
cell_15    1960 non-null int64
cell_16    1960 non-null int64
cell_17    1960 non-null int64
cell_18    1960 non-null int64
cell_19    1960 non-null int64
cell_20    1960 non-null int64
cell_21    1960 non-null int64
cell_22    1960 non-null int64
cell_23    1960 non-null int64
cell_24    1960 non-null int64
cell_25    1960 non-null int64
cell_26    1960 non-null int64
cell_27    1960 non-null int64
cell_28    1960 non-null int64
cell_29    1960 non-null 

## 2 ¿ Qué valores que toma el DataFrame?

Para realizar un análisis adecuado, es muy importante tomar en cuenta el rango de valores que toma cada columna de nuestra tabla, esto nos ayudará más adelante a elegir los métodos más adecuados para realizar análisis cuantitativos.

Ahora exploraremos los valores que toma cada columna. Para ver los valores diferentes que toma la primer columna, utilizaremos lo siguiente:

In [4]:
rawdata['cell_1'].unique()

array([0, 1])

Podemos ver que los valores que toma son **0**, correspondiente a la célula **en reposo** y **1, correspindiente a la célula presentando un potencial de acción**.

Podemos hacer esta misma exploración para el resto de las células:

In [5]:
unique_values = [rawdata[cell].unique() for cell in rawdata.columns]
unique_values

[array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([1, 0]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0]),
 array([0, 1]),
 array([0]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1]),
 array([0, 1])]

## 3 Descriptivos Estadísticos 
Para obtenet los descriptivos generales de un DataFrame, basta con usar el método **.describe()**:

In [6]:
description_df =rawdata.describe()
description_df

cell_1       cell_2       cell_3       cell_4       cell_5  \
count  1960.00000  1960.000000  1960.000000  1960.000000  1960.000000   
mean      0.07551     0.088265     0.007653     0.007143     0.010204   
std       0.26428     0.283753     0.087169     0.084235     0.100524   
min       0.00000     0.000000     0.000000     0.000000     0.000000   
25%       0.00000     0.000000     0.000000     0.000000     0.000000   
50%       0.00000     0.000000     0.000000     0.000000     0.000000   
75%       0.00000     0.000000     0.000000     0.000000     0.000000   
max       1.00000     1.000000     1.000000     1.000000     1.000000   

            cell_6       cell_7       cell_8       cell_9      cell_10  ...  \
count  1960.000000  1960.000000  1960.000000  1960.000000  1960.000000  ...   
mean      0.005102     0.001020     0.003571     0.011735     0.143367  ...   
std       0.071264     0.031936     0.059670     0.107717     0.350537  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

           cell_29      cell_30      cell_31  cell_32      cell_33  \
count  1960.000000  1960.000000  1960.000000   1960.0  1960.000000   
mean      0.014796     0.016327     0.078061      0.0     0.061735   
std       0.120766     0.126760     0.268336      0.0     0.240734   
min       0.000000     0.000000     0.000000      0.0     0.000000   
25%       0.000000     0.000000     0.000000      0.0     0.000000   
50%       0.000000     0.000000     0.000000      0.0     0.000000   
75%       0.000000     0.000000     0.000000      0.0     0.000000   
max       1.000000     1.000000     1.000000      0.0     1.000000   

           cell_34      cell_35      cell_36      cell_37      cell_38  
count  1960.000000  1960.000000  1960.000000  1960.000000  1960.000000  
mean      0.161735     0.004082     0.005612     0.104082     0.006122  
std       0.368301     0.063773     0.074723     0.305444     0.078026  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 38 columns]

## 4 Conteo de Spikes
Para acceder al número de unos y ceros que contiene cada columna es posible usar el método **.value_counts( )** :

In [7]:
rawdata['cell_15'].value_counts()

0    1825
1     135
Name: cell_15, dtype: int64

Podemos ver que la **célula 15** presenta 135 spikes durante el registro.
Obtendremos estos mismos conteos para el resto de las células:

In [8]:
val_counts = [rawdata[cell].value_counts() for cell in rawdata.columns]
val_counts[0] # Conteo de valores para la célula 1.

0    1812
1     148
Name: cell_1, dtype: int64

### 4.1 Tabla resumen de la actividad celular

Ahora crearemos una tabla que incluya el número de 0's y 1's por célula, además del resumen estadístico:

In [9]:
cell_summary = [val_counts[i].append(description_df.iloc[:,i]) for i in range(nCells)]
cell_summary

[0        1812.00000
 1         148.00000
 count    1960.00000
 mean        0.07551
 std         0.26428
 min         0.00000
 25%         0.00000
 50%         0.00000
 75%         0.00000
 max         1.00000
 Name: cell_1, dtype: float64, 0        1787.000000
 1         173.000000
 count    1960.000000
 mean        0.088265
 std         0.283753
 min         0.000000
 25%         0.000000
 50%         0.000000
 75%         0.000000
 max         1.000000
 Name: cell_2, dtype: float64, 0        1945.000000
 1          15.000000
 count    1960.000000
 mean        0.007653
 std         0.087169
 min         0.000000
 25%         0.000000
 50%         0.000000
 75%         0.000000
 max         1.000000
 Name: cell_3, dtype: float64, 0        1946.000000
 1          14.000000
 count    1960.000000
 mean        0.007143
 std         0.084235
 min         0.000000
 25%         0.000000
 50%         0.000000
 75%         0.000000
 max         1.000000
 Name: cell_4, dtype: float64, 0        

### 4.2 Convertir una lista o arreglo a DataFrame
Ahora, convertiremos a DataFrame la lista anterior:

In [10]:
cell_summary_df = pd.DataFrame(cell_summary)
cell_summary_df.head()

0      1   count      mean       std  min  25%  50%  75%  max
cell_1  1812.0  148.0  1960.0  0.075510  0.264280  0.0  0.0  0.0  0.0  1.0
cell_2  1787.0  173.0  1960.0  0.088265  0.283753  0.0  0.0  0.0  0.0  1.0
cell_3  1945.0   15.0  1960.0  0.007653  0.087169  0.0  0.0  0.0  0.0  1.0
cell_4  1946.0   14.0  1960.0  0.007143  0.084235  0.0  0.0  0.0  0.0  1.0
cell_5  1940.0   20.0  1960.0  0.010204  0.100524  0.0  0.0  0.0  0.0  1.0

# 5 Filtrando columnas 
Finalmente, algo que podemos hacer para tener un conjunto de datos más limpio, es eliminar aquellas células que no hayan presentado *spikes* durante el registro.

Para esto, primero contaremos el total de *spikes* que realizó cada célula:

In [11]:
rawdata['cell_1'].sum()

148

In [12]:
#Primero sumamos el total de spikes por cada célula (columna)
spikes_cell = [rawdata[cell].sum() for cell in rawdata.columns]

#... y convertimos la lista anterior a un arreglo:
spikes_cell = np.asarray(spikes_cell)
spikes_cell

array([148, 173,  15,  14,  20,  10,   2,   7,  23, 281, 176, 116,  21,
        16, 135,   0,  20,   0,   2,   6,  81,  91,  76, 148, 101,  62,
       120,   4,  29,  32, 153,   0, 121, 317,   8,  11, 204,  12])

### 5.1 Células Activas e Inactivas

In [13]:
inactiveCells = sc.where(spikes_cell==0)[0]
activeCells = sc.where(spikes_cell>0)[0]

print("El número de células activas es: %s" % (len(activeCells)))
print("El número de células inactivas es: %s" % (len(inactiveCells)))

El número de células activas es: 35
El número de células inactivas es: 3


La lista de células inactivas es la siguiente:

In [14]:
inactiveCells

array([15, 17, 31])

### 5.2 Método .drop( )
Ahora, quitaremos las células inactivas y guardaremos el resultado en la nueva variable *workdata*, para esto utilizaremos el método **.drop( )**.

In [15]:
workdata_active = rawdata.drop(rawdata.columns[inactiveCells], axis=1)
workdata_active.shape

(1960, 35)

In [16]:
workdata_active.to_csv('CellsData_clean.csv')